[PKM-9] feat: Async Redux setup

yohancai opened the pull request:
- created temporary pokemon list page, vm, connector
 - initialized app state
 - created initial UI for the pokemons
 - created tentative persistor

https://user-images.githubusercontent.com/58543326/180947905-a26dc4ed-a6bd-48ea-ad98-cde161932c5b.mp4

Commit:
[PKM-9] feat: Async Redux setup

 - created temporary pokemon list page, vm, connector
 - initialized app state
 - created initial UI for the pokemons
 - created tentative persistor

raelsj committed:
Create lint_workflow.yml

Commit:
lint checker

raelsj committed:
Update lint_workflow.yml

raelsj committed:
Update lint_workflow.yml

raelsj committed:
update workflow

raelsj committed:
Update lint_workflow.yml

raelsj committed:
Update lint_workflow.yml

raelsj committed:
Update lint_workflow.yml

raelsj committed:
Update lint_workflow.yml

raelsj committed:
Update lint_workflow.yml

Commit:
[PKM-9] feat: Async Redux setup

created temporary pokemon list page, vm, connector
initialized app state
created initial UI for the pokemons
created tentative persistor

Commit:
Merge branch 'feature/dev_9' of github.com:shawn-ffufm/pokedex-async-redux into feature/dev_9

Commit:
[PKM-9] feat: Async Redux setup

- Removed lint workflow (being fix by boss rael)
- construct image sized box into 1 line

Commit:
[PKM-9] feat: Async Redux setup

 - Fixed lint via lint commands
 - Deleted generated files

shawn-ffufm commented on lib/features/pokemon_list_connector.dart:
```diff
@@ -0,0 +1,19 @@
+import 'package:async_redux/async_redux.dart';
+import 'package:flutter/material.dart';
+import 'package:pokedex/features/pokemon_list_page.dart';
```
for this feature maybe you can create a new directory for this something like `pokemon_list` => `lib/features/pokemon_list`, so we can keep things organize for the long run 🙂

Then move all inside all your pokemon_list related classes 🙂

shawn-ffufm commented on lib/features/pokemon_list_connector.dart:
```diff
@@ -0,0 +1,19 @@
+  Widget build(BuildContext context) {
+    return StoreConnector<AppState, PokemonListVm>(
+      vm: () => PokemonListVmFactory(),
+      onInit: (store) async => store.dispatch(GetPokemonList()),
```
realign implementation based on my feedback in your used action for this part 🙂

shawn-ffufm commented on lib/features/pokemon_list_page.dart:
```diff
@@ -0,0 +1,38 @@
+  Widget build(BuildContext context) {
+    return Scaffold(
+      appBar: AppBar(
+        title: const Text('Pokedex'),
```
kindly add a new `strings.dart` file inside your `utils` directory for this and make it as a string constant (i.e. `const String pokedexTitle  'Pokedex'`) 🙂

shawn-ffufm commented on lib/features/pokemon_list_page.dart:
```diff
@@ -0,0 +1,38 @@
+              children: <Widget>[
+                /// TODO: show information that was gathered from api
+                Text(
+                  "Lorem ipsum",
```
kindly add this in your strings.dart 🙂

shawn-ffufm commented on lib/features/pokemon_list_page.dart:
```diff
@@ -0,0 +1,38 @@
+                  "Lorem ipsum",
+                  style: Theme.of(context).textTheme.headline5,
+                ),
+                SizedBox(
```
kindly add a todo comment that you will be creating a reusabel widget for image loading 🙂

shawn-ffufm commented on lib/main.dart:
```diff
@@ -1,68 +1,21 @@
-      ),
-    );
-  }
+  runApp(PokedexApp(store: store));
```
do you really need to inject immediately the store here? 

Instead of passing it through, perhaps you can already provide the store by calling `StoreProvider` here and inject your PokedexApp, which will serve as your returned widget for running the app (i.e. `runApp(StoreProvider<AppState>(store: store, child: const PokedexApp()));`) 🙂 

Besides, you'll be setting-up later your go_router config in your PokedexApp, in there you will provide independent routers for you tab pages which will then callout the StoreConnector for further utilization 🙂

shawn-ffufm commented on lib/pokedex_app.dart:
```diff
@@ -0,0 +1,21 @@
+
+  @override
+  Widget build(BuildContext context) {
+    return StoreProvider(
```
see feedback in `main.dart` regarding the usage of StoreProvider 🙂

shawn-ffufm commented on lib/pokedex_app.dart:
```diff
@@ -0,0 +1,21 @@
+  Widget build(BuildContext context) {
+    return StoreProvider(
+      store: store,
+      child: const MaterialApp(
```
kindly add a todo comment that you're going to implement the go_router setup here 🙂

shawn-ffufm commented on lib/state/actions/pokemon_list_actions.dart:
```diff
@@ -0,0 +1,10 @@
+import 'package:async_redux/async_redux.dart';
+import 'package:pokedex/state/app_state.dart';
+
+class GetPokemonList extends ReduxAction<AppState> {
```
For action naming we normally add "Action" word as its suffix therefore maybe you could rename this to `GetPokemonsAction` rather? 🙂 

also, for your action.dart file instead of having a specific name (`pokemon_list_action.dart`), kindly make it a general one rather => `pokemon_actions.dart`

shawn-ffufm replied:
also add documentation above your Action classes and state its purpose why that action is created 🙂

shawn-ffufm commented on lib/state/actions/pokemon_list_actions.dart:
```diff
@@ -0,0 +1,10 @@
+class GetPokemonList extends ReduxAction<AppState> {
+  /// TODO: Make a reduce function that will return 20 pokemons to the state
+  @override
+  Future<AppState> reduce() async {
```
Do you really need to make this as a Future function for this temporary implementation? This is a bit YAGNI for this case 🙈. Basically you only need to make this `Future` once you're asynchronously awaiting for something 🙂 

For further enhancement of your knowledge, kindly research further for the fundamentals of synchronous and asynchronous process 🙂

shawn-ffufm commented on lib/state/app_state.dart:
```diff
@@ -0,0 +1,16 @@
+@freezed
+class AppState with _$AppState {
+  factory AppState({
+    @Default('') @JsonKey(ignore: true) String name,
```
kindly add a todo comment that this is just for testing purposes and and will later be removed once state is finalized 🙂

shawn-ffufm commented on lib/state/app_state.dart:
```diff
@@ -0,0 +1,16 @@
+    @Default('') @JsonKey(ignore: true) String name,
+  }) = _AppState;
+
+  factory AppState.init() => AppState();
```
this seems unused, maybe you can now remove this? 🙂

shawn-ffufm commented on lib/utils/constant.dart:
```diff
@@ -0,0 +1 @@
+const pokeImage = "https://raw.githubusercontent.com/PokeAPI/sprites/master/sprites/pokemon/1.png";
```
can you add TODO comment that this will be removed once dynamic data is used here? 🙂 

also move this string constant to your `strings.dart` and delete this file since this is YAGNI atm 🙈

shawn-ffufm reviewed (changes requested):
@yohancai see pr feedback 🙂

Commit:
[PKM-9] feat: Async Redux setup

- Updated pokemon list directory
- Updated naming conventions
- Transfered Storeprovider on main
- Created a string.dart file

Commit:
[PKM-9] feat: Async Redux setup

- Updated pokemon list directory
- Updated naming conventions
- Transfered Storeprovider on main
- Created a string.dart file

shawn-ffufm commented on lib/pokedex_app.dart:
```diff
@@ -0,0 +1,16 @@
+  @override
+  Widget build(BuildContext context) {
+    return const MaterialApp(
+        title: 'Pokedex',
```
kindly reuse the string you added in your `strings.dart` 🙂

shawn-ffufm commented on lib/state/actions/pokemon_actions.dart:
```diff
@@ -0,0 +1,11 @@
+import 'package:async_redux/async_redux.dart';
+import 'package:pokedex/state/app_state.dart';
+
+/// This action will be used to generate 20 Pokemon at a time.
```
this will later be used for getting pokemons right? perhaps you could still simply this by still having a todo comment of => `/// todo: documentation will be changed later -> Get paginated pokemons per 20 items`

shawn-ffufm reviewed (changes requested):
@yohancai see last minor pr feedback 🙂

shawn-ffufm commented on lib/features/pokemon_list/pokemon_list_page.dart:
```diff
@@ -0,0 +1,39 @@
+import 'package:pokedex/utils/strings.dart';
+
+class PokemonListPage extends StatelessWidget {
+  const PokemonListPage({super.key, required this.name});
```
add extra trailing comma 🙂

Commit:
[PKM-9] feat: Async Redux setup

- added trailing comma
- Provided appropriate todos

Commit:
[PKM-9] feat: Async Redux setup

- fix appropriate use of string.dart

shawn-ffufm reviewed (approved):
Lgtm 💪

raelsj commented on test/widget_test.dart:
```diff
@@ -1,19 +1,17 @@
-// This is a basic Flutter widget test.
```
you can remove this file

Commit:
[PKM-9] feat: Async Redux setup

- deleted widget_test.dart

raelsj commented on lib/features/pokemon_list/pokemon_list_page.dart:
```diff
@@ -0,0 +1,43 @@
+  Widget build(BuildContext context) {
+    return Scaffold(
+      appBar: AppBar(
+        title: const Text(str.pokedexTitle),
```
remove this trailing comma it's unnecessary

raelsj replied:
@yohancai

raelsj commented on lib/features/pokemon_list/pokemon_list_page.dart:
```diff
@@ -0,0 +1,43 @@
+
+                /// TODO: create a reusable widget for loading the image
+                SizedBox(
+                  width: 100,
```
this 100 have this as a constant

Commit:
[PKM-9] feat: Async Redux setup

- removed unnecessary trailing comma
- created constant variables

raelsj commented on lib/main.dart:
```diff
@@ -1,68 +1,21 @@
-      ),
-    );
-  }
+  runApp(StoreProvider<AppState>(store: store, child: const PokedexApp()));
```
trailing comma here

Commit:
[PKM-9] feat: Async Redux setup

- added trailing comma

raelsj commented on lib/main.dart:
```diff
@@ -1,68 +1,21 @@
-      ),
-    );
-  }
+  runApp(StoreProvider<AppState>(store: store, child: const PokedexApp(),));
```
then format

Commit:
[PKM-9] feat: Async Redux setup

- removed unnecessary trailing comma

raelsj commented on lib/features/pokemon_list/pokemon_list_page.dart:
```diff
@@ -0,0 +1,44 @@
+  Widget build(BuildContext context) {
+    return Scaffold(
+      appBar: AppBar(
+        title: const Text(str.pokedexTitle)
```
then format. always format everything

Commit:
[PKM-9] feat: Async Redux setup

- fix format
